lib/puppetx/simp/compliance_mapper.rb

```diff
@@ -0,0 +1,153 @@
+  case key
+  when "lookup_options"
+    # XXX ToDo See note about compiling a lookup_options hash in the compiler
+    throw :no_such_key
+  when "compliance_map"
+    throw :no_such_key
+  when "compliance_markup::compliance_map"
+    throw :no_such_key
+  when "compliance_markup::compliance_map::percent_sign"
+    throw :no_such_key
+  when "compliance_markup::enforcement"
+    throw :no_such_key
+  when "compliance_markup::version"
+    throw :no_such_key
+  when "compliance_markup::percent_sign"
+    throw :no_such_key
+  else
+    retval = :notfound
+    if cache_has_key(:lock)
+      lock = cached_value(:lock)
+    else
+      lock = false
+    end
+    if (lock == false)
+      cache(:lock, true)
+      begin
+        profile_list = cached_lookup "compliance_markup::enforcement", [], &block
+        unless (profile_list == [])
+          debug("compliance_markup::enforcement set to #{profile_list}, attempting to enforce")
+          version = cached_lookup "compliance_markup::version", "1.0.0", &block
+          case version
+          when /1.*/
+            v1_compliance_map = {}
+
+            if (cache_has_key(:v1_compliance_map))
+              v1_compliance_map = cached_value(:v1_compliance_map)
+            else
+              debug("loading compliance_map data from compliance_markup::compliance_map")
+              module_scope_compliance_map = cached_lookup "compliance_markup::compliance_map", {}, &block
+              top_scope_compliance_map = cached_lookup "compliance_map", {}, &block
+              v1_compliance_map.merge!(module_scope_compliance_map)
+              v1_compliance_map.merge!(top_scope_compliance_map)
+              cache(:v1_compliance_map, v1_compliance_map)
+              # XXX ToDo: Add a dynamic loader for compliance data, so that modules can embed
+              # their own compliance map information. Dylan has a way to do this in testing
+              # in Abacus
+            end
+
+            # Pre-compile the values for each profile list array.
+            # We use hash.to_s, then create a hash named that in v1_compliance_map,
+            # that is a raw key => value mapping. This simplifies our code as we can assume
+            # that if the key exists, then the value is what we use. We also don't have to worry
+            # about exponential time issues since this is linearlly done once, not every time for
+            # every key.
+
+            profile = profile_list.hash.to_s
+            v1_compile(profile, profile_list, v1_compliance_map)
+            if (v1_compliance_map.key?(profile))
+              # Handle a knockout prefix
+              unless (v1_compliance_map[profile].key?("--" + key))
+                if (v1_compliance_map[profile].key?(key))
+                  retval = v1_compliance_map[profile][key]
+                end
+              end
+            end
+          end
+        end
+      ensure
+        cache(:lock, false)
+      end
+    end
+    if (retval == :notfound)
```

Comment:
How does cache get locked?  Do we have multi-threading going on?  What is the implication of returning  not found when locked?  Does this cause the compile to fail?  Can we  wait (with a timeout) to retrieve the cache instead of instantly failing?